### DISCUSSION TITLE
Handling Data Preservation on dora stop and dora destroy Commands in Rust

### DISCUSSION THREAD

**Discussion by Geoffrey-Hooton** [dora-rs/dora#586] [Q&A] [answered]
Hi!

Currently, I have a Dora node written in Rust that is storing sensor data into an HDF5 file every minute or so. I was wondering if there was an easy way to handle a `dora destroy` or `dora stop` command such that, when one of these commands is called, I can execute one last function to store the data I have collected between the last hdf5 file being created and the shutdown, ensuring it is not all lost.

Obviously, in Rust, there are crates like the ctrlc crate to handle interruptions like this. However, we do not use ctrl+c to end a Dora dataflow.

I'm not entirely sure what happens when a `dora stop` or `dora destroy` command is executed, so any clarification/guidance you can provide would be a great help!

Thanks!

**Comment by haixuanTao** [comment, marked as answer]
Hi, thanks for reaching out!

So basically if you use the cli and send a stop signal.

Each node will receive a stop event before. Closing the event stream, that will make the loop return None.

In which case you can write on disk.

You can use --grace-duration if you're not is killed to quickly and does not have time to write on disk.

Does this work for you or do you need a more specific control flow for your dataflow ?

**Reply by Geoffrey-Hooton** [reply]
Yep that worked perfectly! Thanks so much for your help.
